File: src/algorithms/cpp/max-heap.cpp
// Max Heap Insertion
#include <iostream>
#include <vector>

using namespace std;

void insert(vector<int>& heap, int val) {
    heap.push_back(val);
    int curr = heap.size() - 1;
    
    // Sift up
    while (curr > 0) {
        int parent = (curr - 1) / 2;
        if (heap[curr] > heap[parent]) {
            swap(heap[curr], heap[parent]);
            curr = parent;
        } else {
            break;
        }
    }
}

int main() {
    vector<int> arr = {15, 30, 20, 45, 10, 50};
    vector<int> heap;
    
    for (int val : arr) {
        insert(heap, val);
    }
    
    return 0;
}

File: src/algorithms/python/max-heap.py
# Max Heap Insertion
# A Max Heap is a complete binary tree where the parent is always greater than its children.

def insert(heap, val):
    heap.append(val)
    curr = len(heap) - 1
    
    # Sift up
    while curr > 0:
        parent = (curr - 1) // 2
        if heap[curr] > heap[parent]:
            heap[curr], heap[parent] = heap[parent], heap[curr]
            curr = parent
        else:
            break

arr = [15, 30, 20, 45, 10, 50]
heap = []

for val in arr:
    insert(heap, val)
